source/client.c

```diff
@@ -257,6 +258,21 @@ static void s_mqtt_client_init(
         return;
     }
 
+    /* user request disconnect before the channel setted up, stop installing the slot and sending CONNECT. */
+    bool is_closing = false;
+
+    { /* BEGIN CRITICAL SECTION */
+        s_mqtt_connection_lock_synced_data(connection);
+        is_closing = connection->synced_data.state == AWS_MQTT_CLIENT_STATE_DISCONNECTING;
+        s_mqtt_connection_unlock_synced_data(connection);
+    } /* END CRITICAL SECTION */
+    if (is_closing) {
```

Comment:
to be honest, I don't see what good this does.  the client could easily be set to disconnecting right after this check is made.

Reply:
The point is the `mqtt_disconnect_impl` will do nothing if the slot is not installed. As long as the state is set to disconnecting (`mqtt_disconnect_impl` happens) after the slot is installed, we will be fine.

Reply:
or just extend the lock until then maybe?

Reply:
Yeah, this is the only confusing bit left, what is the rationale here?

Reply:
The reason for this early stop here is that if the slot is not installed and user request to disconnect, the connection will not be closed from our side. We will wait for the broker/server side to close the connection, which is different from what user wants. 
So, we need to ensure the disconnect happens before the slot installed will fall into the early stop and connection is closed from our side as user requested.